### PR TITLE
wip: w3capi in browser

### PR DESCRIPTION
**Pull request by lisalove**

### CODE REVIEW

File: js/generate.js [hunk 1]
```diff
@@ -38,23 +38,24 @@ function fillCell(el, data, image) {
...
-	    var svg = e.target.response.documentElement;
-	    svg.querySelector("style").remove();
-	    el.appendChild(svg);
-	    if (postHook) {
-		postHook(el);
-	    }
-	}
-    };
-    xhr.send();
+    fetch(svgurl, {mode:'cors'}).then(response => response.text())
```
Comment (marcoscaceres) [inline js/generate.js:41]: So, worried that this will break iOS and MS Edge, etc. that don't yet support fetch().

File: js/generate.js [hunk 2]
```diff
@@ -38,23 +38,24 @@ function fillCell(el, data, image) {
...
+        .then(function(svg) {
+            if (svg) {
+                var d = document.createElement("p");
```
Comment (marcoscaceres) [inline js/generate.js:44]: what's the rationale of putting the SVG into a p element? Would a figure not be more appropriate?

File: js/generate.js [hunk 3]
```diff
@@ -38,23 +38,24 @@ function fillCell(el, data, image) {
...
+                d.innerHTML = svg;
+                d.querySelector("style").remove();
+                el.appendChild(d);
+                if (postHook) {
+                    postHook(el);
+                }
+            }
+        }).catch(function(e) {
+            console.error(e);
```
Comment (marcoscaceres) [inline js/generate.js:53]: here you use function(), but below you use a console.xxx.bind(). Would be great if the two bound functions could just be declared a variable:

``` JS
const consoleErr = console.error.bind(console);
...
.catch(consoleErr);
```

Or some such...

File: js/generate.js [hunk 4]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
 	maturity = {label: spec.maturity, level:level};
     } else {
 	maturity = {label:spec.maturity, level: maturityLevels[spec.maturity]};
-	maturityIcon = {src:"http://www.w3.org/2013/09/wpd-rectrack-icons/" + spec.maturity.toLowerCase().replace(/lastcall/,'lcwd') + '.svg', alt:spec.maturity, width:50, height:50};
-	if (spec.maturity == "REC" || spec.maturity == "LastCall") {
+	maturityIcon = {src:"http://www.w3.org/2013/09/wpd-rectrack-icons/" + maturityMapping[spec.maturity].toLowerCase() + '.svg', alt:spec.maturity, width:50, height:50};
+	if (spec.maturity == "Recommendation" || spec.maturity == "Last Call") {
```
Comment (marcoscaceres) [inline js/generate.js:71]: nit: ===

File: js/generate.js [hunk 5]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
-var specXhr = new XMLHttpRequest();
-specXhr.open("GET", "specs/tr.json");
-specXhr.onload = function() {
-    specData = JSON.parse(this.responseText);
-    fillTables();
-};
-specXhr.send();
+var specData = {};
+fillTables();
+var w3cApiAuth = "?apikey=rvl38z1hpvkw0g804gk48o0s8408w88";
```
Comment (marcoscaceres) [inline js/generate.js:80]: seems funny these keys are so public. I guess it doesn't matter :dancer:

File: js/generate.js [hunk 6]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+var w3cApiFetchConfig = { method:'GET', mode:'cors'};
+
+
+function fetchSpecData(d) {
```
Comment (marcoscaceres) [inline js/generate.js:84]: Suggestion, you could destructure `d` maybe? It would make it more clear:

``` JS
function fetchSpecData({maturity, wgs, title, TR} = {}) {
    if (!TR) {
        return new Promise((res,rej) => { res({maturity, wgs, title});});
    }
```

Comment (marcoscaceres) [inline js/generate.js:84]: Also, rej, is defined by never used.

File: js/generate.js [hunk 7]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+    if (!d.TR) {
+        return new Promise((res,rej) => { res({maturity:d.maturity, wgs: d.wgs, title: d.title});});
+    }
+    var shortName = d.TR.split('#')[0].split('/')[4];
+    var ret = {};
+    return fetch("https://api.w3.org/specifications/" + shortName + "/versions" + w3cApiAuth + "&embed=1", w3cApiFetchConfig)
+        .then(response =>  response.json())
```
Comment (marcoscaceres) [inline js/generate.js:91]: nit, whitespace

File: js/generate.js [hunk 8]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+        .then(function(data) {
```
Comment (marcoscaceres) [inline js/generate.js:92]: nit: use fat arrow

File: js/generate.js [hunk 9]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+            var latest = data._embedded['version-history'][0];
+            ret = { maturity: latest.status, title: latest.title};
```
Comment (marcoscaceres) [inline js/generate.js:94]: You are trashing your original ret here. Why not just override the properties or properties? Also, you could use Object.assign() instead.

File: js/generate.js [hunk 10]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+            return fetch(latest._links["deliverers"].href  + w3cApiAuth + "&embed=1", w3cApiFetchConfig).then(response => response.json());
```
Comment (marcoscaceres) [inline js/generate.js:95]: nit: Break at .then

File: js/generate.js [hunk 11]
```diff
@@ -66,22 +67,38 @@ function maturityData(spec) {
...
+        })
+        .then(function(data) {
+            ret["wgs"] = data._embedded.deliverers.map(g => ({label: g.name, url: g._links["homepage"].href}));
```
Comment (marcoscaceres) [inline js/generate.js:98]: Nit: you are accessing a "_private" prop? Maybe expose it as public on whatever g is?

File: js/generate.js [hunk 12]
```diff
@@ -140,18 +158,10 @@ function fillTables() {
...
 		var docTd = document.createElement("td");
 		var tsTd = document.createElement("td");
-		var xhr = new XMLHttpRequest();
-		xhr.open("GET", "data/" + spec + ".json");
-                counterReq++;
-		xhr.onload = function(x, s, el1, el2, el3, el4, el5, el6, el7, el8) {
-		    return function() {
-                        counterRes++;
+                
+		function trClosure(s, el1, el2, el3, el4, el5, el6, el7, el8) {
```
Comment (marcoscaceres) [inline js/generate.js:162]: :heart: ... but :open_mouth:!? If you are going ES6 here, then maybe use `...elems` as a rest parameter. Then you can access them in array order. 

``` JS
function trClosure(s, ...elems) {
}
```

**Comment by marcoscaceres** [comment]
this code might benefit a bit from a run over JSCS and JSHint or ESLint.

**Comment by lisalove** [comment]
Thank you so much for your help with this, I am just learning about this stuff and it's kinda driving me crazy trying to figure it all out.

Sent by Outlookhttp://taps.io/outlookmobile for Android

On Sun, Feb 14, 2016 at 5:36 PM -0800, "Marcos Caceres" <notifications@github.com<mailto:notifications@github.com>> wrote:

this code might benefit a bit from a run over JSCS and JSHint or ESLint.

##

**Comment by marcoscaceres** [comment]
@lisalove, feel free to ask questions. We are happy to help if you get stuck or if you need any guidance.

File: js/generate.js [hunk 13]
```diff
@@ -214,13 +221,19 @@ function fillTables() {
 			}
 			fillCell(el8, data.tests);
 			el8.classList.add("tests");
-                        if (counterReq == counterRes) {
+		        });
+                    };
+                }
+                var p = fetch("data/" + spec + ".json").then(
```
Comment (marcoscaceres) [inline js/generate.js:227]: nit, break on .then() here too.

Comment (lisalove) [inline js/generate.js:227]: Thank you so much Marcos. What does the text say?

Comment (marcoscaceres) [inline js/generate.js:227]: > What does the text say?

Sorry, not sure what you are are asking here. 

Before proceeding, we should really wait for @dontcallmedom to comment on the use of `fetch()`. If code is to be compatible with today's browsers (or if this code is tested with PhantomJS), then using `fetch()` might be a show-stopper.

**Comment by dontcallmedom** [comment]
@marcoscaceres to clarify I was the one who wrote the patch that @lisalove submitted as a pull request; as its commit title, it is (or was) a work in progress; I have since left it aside: since the w3C API doesn't provide caching headers, and since it also makes it hard to follow some redirects, I've put this in the backburner.

(in practice, the reliance on fetch would not be an issue in the current usage, since basically the JS is "only" used when generating the static final document, which can be assumed to be done in a modern browser)

I'm closing this pull request since I see no reason to merge an unfinished work I started :)
